NetCasbin/Model/Assertion.cs

```diff
@@ -16,29 +17,33 @@ public class Assertion
 
         public string[] Tokens { set; get; }
 
-        private List<List<string>> _policy;
+        public IRoleManager RoleManager { get; private set; }
 
-        private IRoleManager _rm;
+        public List<List<string>> Policy { set; get; }
 
-        public IRoleManager RM => _rm;
+        internal HashSet<string> PolicyStringSet { get; }
 
-        public List<List<string>> Policy
+        public Assertion()
         {
-            set => _policy = value;
-            get => _policy;
+            Policy = new List<List<string>>();
+            PolicyStringSet = new HashSet<string>();
```

Comment:
The `PolicyStringSet` just used to store rule? what if a rule contains same string, like

p, alice, alice, read

Reply:
It stores the string which connected by rule `List<string>`, I think it will keep the same operation as before.
this is operation before :
```csharp
for (var i = 0; i < a.Count; i++)
{
    if (!a[i].Equals(b[i]))
    {
        return false;
    }
}
return true;
```
If a rule contains the same string and the `HasPolicy` will can not return false at before too.

Reply:
Ok, understand, you turn an array into a string, so no need to worry about this.

Reply:
Why we don't use `HashSet<List<string>`? somethings like:
```c# 
 class ListPolicy<T> : List<T>
    {
        public override bool Equals(object obj)
        {
           // I'm not sure how to write.
            return obj is ListPolicy<T> policy &&
                   Capacity == policy.Capacity &&
                   Count == policy.Count;
        }

        public override int GetHashCode()
        {
            var hash = new HashCode();
            foreach (var item in this)
            {
                hash.Add(item);
            }
            return hash.ToHashCode();
        }
    }



    class MainClass
    {
        public static void Main(string[] args)
        {

            var policy = new HashSet<ListPolicy<string>>
            {
                new ListPolicy<string> { "user","data1","read" },
                new ListPolicy<string> { "user","data1","read" }
            };

            Console.WriteLine(policy.Count); // => 1
        }
    }
```

Reply:
> Why we don't use `HashSet<List<string>`? somethings like:
> 
> ```cs
>  class ListPolicy<T> : List<T>
>     {
>         public override bool Equals(object obj)
>         {
>            // I'm not sure how to write.
>             return obj is ListPolicy<T> policy &&
>                    Capacity == policy.Capacity &&
>                    Count == policy.Count;
>         }
> 
>         public override int GetHashCode()
>         {
>             var hash = new HashCode();
>             foreach (var item in this)
>             {
>                 hash.Add(item);
>             }
>             return hash.ToHashCode();
>         }
>     }
> 
> 
> 
>     class MainClass
>     {
>         public static void Main(string[] args)
>         {
> 
>             var policy = new HashSet<ListPolicy<string>>
>             {
>                 new ListPolicy<string> { "user","data1","read" },
>                 new ListPolicy<string> { "user","data1","read" }
>             };
> 
>             Console.WriteLine(policy.Count); // => 1
>         }
>     }
> ```

I mentioned it in the next review. here is the content:

If changed `List<List<string>>` to `Hash<Rule>` (the `Rule` inherit `List<string>` and overwrite the `Equals` and `GetHashCode`), It will change almost all the public API and can not gain more performance. It will cost much time when `Add` the list because the `Equal` or `Add` has  O(n) complexity.

Additionally, The `HashSet` can not get by index and will shuffle the list (But we can add the index property to handle it).

Reply:
@nodece @xcaptain Here is the benchmark result of using `HashSet<Rule>`, the code is here [change-to-hashset branch](https://github.com/Sagilio/Casbin.NET/tree/change-to-hashset) :
``` ini

BenchmarkDotNet=v0.12.1, OS=Windows 10.0.19645
Intel Core i7-7700HQ CPU 2.80GHz (Kaby Lake), 1 CPU, 8 logical and 4 physical cores
.NET Core SDK=5.0.100-preview.5.20279.10
  [Host]     : .NET Core 3.1.2 (CoreCLR 4.700.20.6602, CoreFX 4.700.20.6702), X64 RyuJIT
  Job-WRAGMF : .NET Core 3.1.2 (CoreCLR 4.700.20.6602, CoreFX 4.700.20.6702), X64 RyuJIT

IterationCount=3  RunStrategy=Throughput  

```
|    Method | Times | PolicyCount |          Mean |         Error |      StdDev |           Min |           Max |        Median |
|---------- |------ |------------ |--------------:|--------------:|------------:|--------------:|--------------:|--------------:|
| **HasPolicy** |  **1000** |         **500** |    **511.904 μs** |   **331.9158 μs** |  **18.1934 μs** |    **492.169 μs** |    **528.009 μs** |    **515.534 μs** |
| AddPolicy |  1000 |         500 | 10,286.494 μs | 1,392.4007 μs |  76.3222 μs | 10,239.692 μs | 10,374.566 μs | 10,245.225 μs |
| **HasPolicy** |  **1000** |        **5000** |    **523.930 μs** |   **411.9189 μs** |  **22.5787 μs** |    **498.292 μs** |    **540.851 μs** |    **532.646 μs** |
| AddPolicy |  1000 |        5000 | 10,284.017 μs | 5,746.3454 μs | 314.9766 μs |  9,943.827 μs | 10,565.527 μs | 10,342.697 μs |